### DISCUSSION TITLE
Original file path from `SourceLocation`

### DISCUSSION THREAD

**Discussion by fabianschuiki** [MikePopoloski/slang#658] [Q&A]
I'm experimenting with integrating slang into CIRCT, which is built around MLIR. Location information in MLIR diagnostics is expected to produce the exact path names as given by the user, without any modification.

Right now I'm using `sourceManager.getFileName(loc)` to get the path to report in the MLIR diagnostic. That function though returns the file _name_, which is either the proximate path or -- if `disableProximatePaths` is on -- just the file name, but not the actual path itself.

Is there any way to get the actual _path_ as originally provided when populating the source manager? If not, could I contribute a PR that introduces a `sourceManager.getFilePath(loc)` function such that you can get the full path?

Thanks!

**Comment by MikePopoloski** [comment]
I don't mind a PR if you think it's necessary, but is it not sufficient to do something like call std::filesystem::canonical() on the result of getFileName? If not, or you want to avoid the extra call because you're doing that query a lot, you can store the absolute path in the FileData object directly, since we already know it in openCached at the time we do the read call.

Also of note is that getFileName takes into account `line directives. I'm not sure how you want to handle those, but if you want to ignore them then you should use getRawFileName instead.

**Reply by MikePopoloski** [reply]
Or after reading your message again, if you want exactly whatever the user provided to the tool then disableProximatePaths is almost what you want, it's just that cacheBuffer is chopping it down to just the filename when it could instead take the whole string. disableProximatePaths is currently only used by unit tests so you can feel free to change it however you want to better suit your purposes.

**Reply by fabianschuiki** [reply]
The line directives are an excellent point. Not sure how to deal with those. I guess the user expectation would be to just produce those verbatim in the diagnostic.

Storing the path in the `FileInfo` and adding a way to retrieve that sounds like a good plan. Reproducing the exact thing that was given to the initial source buffer construction sounds like something that might be interesting to other users of slang as well.
